scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala

```diff
@@ -2146,6 +2153,36 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
     }
   }
 
+  private def followingIsLambdaAfterColon(): Boolean = {
+    /** Skip matching pairs of `(...)` or `[...]` parentheses.
+     *  @pre  The current token is `(` or `[`
+     */
+    def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
```

Comment:
slightly rewritten, let me know if it's more readable:
```
def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
  // starts on left delimiter
  @tailrec
  def iter = {
    next()
    if (token.is[EOF]) false
    else if (token.is[T2]) true
    else  (token.isNot[T1] || skipParens[T1, T2]) && iter
  }
  // stops on right delimiter (true), or end (false)
  iter
}

ahead {
  val okParams = token match {
    case _: Ident | _: Underscore => true
    case _: LeftParen => skipParens[LeftParen, RightParen]
    case _: LeftBracket => skipParens[LeftBracket, RightBracket]
    case _ => false
  }
  okParams && {
    next()
    token.is[RightArrow] || token.is[ContextArrow]
  } && {
    next()
    token.is[Indentation.Indent]
  }
}
```

Reply:
Thanks! I looks better, I already rewrote it a bit from the Dotty version, but it was still lacking :sweat_smile:

Reply:
could also do this, perhaps, without skipParens recursion:
```
def skipParens[T1: TokenClassifier, T2: TokenClassifier]: Unit = {
  // starts on left delimiter
  @tailrec
  def iter(nest: Int) = {
    next()
    if (token.is[EOF]) false
    else if (token.is[T2]) nest == 0 || iter(nest - 1)
    else (token.is[T1]) iter(nest + 1)
    else iter(nest)
  }
  // stops on right delimiter (true), or end (false)
  iter(0)
}
```

Reply:
Good idea, I also added a test which has multiple `(` and `[`. I had to slightly rewrite TreeSyntax since it would always change to `{` if type was defined for parameter, but it's actually not needed since we already wrap it in `()` for all cases beside implicit and using